x-pack/plugins/data_enhanced/public/search/search_abort_controller.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { NEVER, Subscription, timer } from 'rxjs';
+
+export enum AbortReason {
+  Timeout = 'timeout',
+}
+
+export class SearchAbortController {
+  private inputAbortSignals: AbortSignal[] = new Array();
+  private abortController: AbortController = new AbortController();
+  private timeoutSub?: Subscription;
+  private boundAbortHandler: (this: AbortSignal, ev: Event) => void;
+  private destroyed = false;
+  private reason?: AbortReason;
+
+  constructor(abortSignal?: AbortSignal, timeout?: number) {
+    this.boundAbortHandler = this.abortHandler.bind(this);
+    if (abortSignal) {
+      this.addAbortSignal(abortSignal);
+    }
+
+    if (timeout) {
+      const timeout$ = timeout ? timer(timeout) : NEVER;
```

Comment:
nit: we've already checked for `if(timeout) a line above. no need for conditions here